demo/src/features/debug3d/mod.rs

```diff
@@ -1,53 +1,71 @@
-use crate::features::debug3d::extract::Debug3dExtractJob;
-use rafx::framework::{VertexDataLayout, VertexDataSetLayout};
-use rafx::nodes::ExtractJob;
-use rafx::nodes::RenderFeature;
-use rafx::nodes::RenderFeatureIndex;
-use std::convert::TryInto;
-
-mod debug3d_resource;
+rafx::declare_render_feature_mod!();
+rafx::declare_render_feature_renderer_plugin!();
+
+rafx::declare_render_feature!(Debug3DRenderFeature, DEBUG_3D_FEATURE_INDEX);
+
 mod extract;
-mod plugin;
 mod prepare;
 mod write;
-pub use plugin::Debug3DRendererPlugin;
 
-pub use debug3d_resource::*;
-use rafx::api::RafxPrimitiveTopology;
+mod public;
 
-pub fn create_debug3d_extract_job() -> Box<dyn ExtractJob> {
-    Box::new(Debug3dExtractJob::new())
-}
+use distill::loader::handle::Handle;
+use rafx::assets::MaterialAsset;
 
-pub type Debug3dUniformBufferObject = shaders::debug_vert::PerFrameUboUniform;
+pub use public::debug3d_resource::DebugDraw3DResource;
 
-/// Vertex format for vertices sent to the GPU
-#[derive(Clone, Debug, Copy, Default)]
-#[repr(C)]
-pub struct Debug3dVertex {
-    pub pos: [f32; 3],
-    pub color: [f32; 4],
+struct StaticResources {
+    pub debug3d_material: Handle<MaterialAsset>,
 }
 
-lazy_static::lazy_static! {
-    pub static ref DEBUG_VERTEX_LAYOUT : VertexDataSetLayout = {
-        use rafx::api::RafxFormat;
+pub struct RendererPluginImpl;
+
+impl RendererPlugin for RendererPluginImpl {
+    fn configure_render_registry(
+        &self,
```

Comment:
I preferred the plugin in a separate file

Reply:
I consolidated the plugin / feature definition into the `mod` file because it provides a single location to see the following:

1. The `public` feature declaration itself, e.g. `rafx::declare_render_feature!(ImGuiRenderFeature, DEBUG_3D_FEATURE_INDEX);`
2. Any `public` structs that have been re-exported, e.g. `pub use public::Sdl2ImguiManager;`
3. The static resources the feature expects.
4. The `public` plugin itself, which is really the _main_ interface, and defines the `init` & `destroy` code. (possibly referencing the internal or public structs from 2 & shoving them into `Resources`), including the `initialization` of the static resources from 3.

Reply:
When a mod.rs file gets to be 100+ lines, I start looking for ways to pull larger chunks of code out of it. The mesh plugin is already close to 100 lines by itself. In general the plugin seems like a piece that could grow. (pub(super) is very helpful for this, allowing something to act like it's in the mod.rs file for visibility purposes but not having to actually be in the mod.rs) Examples of this are in asset_manager.rs, resource_manager.rs, and descriptor_set_allocator.rs. They are the "top-level" interfaces for the "folder" they are in. The mod.rs in those cases contains small misc. bits or the using statements. I think 1, 2, and 3 are fine to be in the top level mod file. I think 3 is fine to be stored along with 4 (as it is now) because in *most* cases it's not a public-facing struct. However, it's likely to be a small type so I don't have a strong opinion on it.

Reply:
See reply here. https://github.com/aclysma/rafx/pull/122#discussion_r610996877